Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+* `runtime_config_path` - path to the `.runtimeconfig.json` file to process. Unlike with the `hostfxr_initialize_for_app`, any `.deps.json` from the app/component will not be processed by the hosting layers.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `host_context_handle` - output parameter. On success receives an opaque value which identifies the initialized host context. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can be called multiple times in a process.
+* If it's called when no runtime is present, it will run through the steps to "initialize" the runtime (resolving frameworks and so on).
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
+
+It needs to be possible to call this function simultaneously from multiple threads at the same time.
+It also needs to be possible to call this function while there is an active host context created by `hostfxr_initialize_for_app` and running inside the `hostfxr_run_app`.
+
+The function returns specific return code for the first initialized host context, and a different one for any subsequent one. Both return codes are considered "success". If there already was initialized host context in the process then the returned host context has these limitations:
+* It won't allow setting runtime properties.
+* The initialization will compare the runtime properties from the `.runtimeconfig.json` specified in the `runtime_config_path` with those already set to the runtime in the process
+  * If all properties from the new runtime config are already set and have the exact same values (case sensitive string comparison), the initialization succeeds with no additional consequences. (Note that this is the most typical case where the runtime config have no properties in it.)
+  * If there are either new properties which are not set in the runtime or ones which have different values, the initialization will return a special return code - a "warning". It's not a full on failure as initialized context will be returned.
+  * In both cases only the properties specified by the new runtime config will be reported on the host context. This is to allow the native host to decide in the "warning" case if it's OK to let the component run or not.
+  * In both cases the returned host context can still be used to get a runtime delegate, the properties from the new runtime config will be ignored (as there's no way to modify those in the runtime).
+
+
+### Inspect and modify host context
+
+#### Runtime properties
+These functions allow the native host to inspect and modify runtime properties.
```

Comment:
I was reading through the explanation for these helper functions...
hostfxr_get_runtime_property
hostfxr_set_runtime_property
hostfxr_get_runtime_property_count
hostfxr_get_runtime_property_name

I can see how to use them, but I am wondering if an alternative way of allowing update was considered.  Is it possible to return the key/value pairs and that would be mutable?

Reply:
Mutable structs in C/C++ across module boundaries are tricky due to memory ownership. If the caller would set a new value, the buffer with the value would have to come from the caller, and thus it would have to somehow guarantee its lifetime and so on...

The other complication is that givin access to structs gives up any control over what is allowed when. For example in some cases setting property values is not supported (if the runtime is already running) - in such case the host would have to somehow detect changes and fail, which is complicated. It also doesn't fail at the place where the modification is made.

I will change these per @AaronRobinsonMSFT  suggestion so that they should be easier to use.

Reply:
Once the changes are pushed, I will take a look.

Reply:
They already are in the latest version.